### PR TITLE
Poq issue 1409 log

### PR DESCRIPTION
**Pull request by poquirion**
We should document how to tweek  log level int the ~/.sctrc file

I have called the logger "sct"

### Description of the Change

Now `sct.printv` is used in most place

`sct.printv` uses `logger.info()` instead of `print()`

### Applicable Issues

Implements #1409

### CODE REVIEW

**Comment by poquirion** [comment]
@jcohenadad @peristeri 

I am pretty happy with where this is going. There is still a couple places in the code where we should use the logger, but I do not think it prevent us from integrating logs right now.

**Comment by jcohenadad** [comment]
i'd like to merge https://github.com/neuropoly/spinalcordtoolbox/pull/1417 and do a few tests before merging this PR

**Comment by jcohenadad** [comment]
verbose problem (to reproduce, simply launch propseg):
~~~
Remove temporary files...
fatal: Not a git repository (or any of the parent directories): .git
Sucessfully generated the QC results in /Users/julien/qc_data/qc_results.json
INFO:sct:Sucessfully generated the QC results in /Users/julien/qc_data/qc_results.json
Use the following command to see the results in a browser:
INFO:sct:Use the following command to see the results in a browser:
sct_qc -folder /Users/julien/qc_data
INFO:sct:sct_qc -folder /Users/julien/qc_data

Done! To view results, type:
INFO:sct:
Done! To view results, type:
fslview t1.nii.gz t1_seg.nii.gz -l Red -b 0,1 -t 0.7 &

INFO:sct:fslview t1.nii.gz t1_seg.nii.gz -l Red -b 0,1 -t 0.7 &
~~~

**Comment by poquirion** [comment]
For sct_pipeline, the integration is note complete, that is right.... I'll have a look at the propseg problem asap.

**Comment by poquirion** [comment]
@jcohenadad 
Just to be sure the problems are only with `sct_pipeline`, because I don't see any problems with the tools themself. I run `sct_propseg` and do not see problems.

**Comment by poquirion** [comment]
For Cedar:
Its the mpi version that is not good anymore We need >2.0 and I don't know why the installer now get the 1.4.1 version. 

:confused:

**Comment by jcohenadad** [comment]
@poquirion: the problem of empty log is indeed with sct_pipeline only. However all tools have a verbose problem: they work, but the verbose is weird: i did a copy/paste in my comment above-- here again:
~~~
INFO:sct:Use the following command to see the results in a browser:
sct_qc -folder /Users/julien/qc_data
INFO:sct:sct_qc -folder /Users/julien/qc_data

Done! To view results, type:
INFO:sct:
Done! To view results, type:
fslview t1.nii.gz t1_seg.nii.gz -l Red -b 0,1 -t 0.7 &

INFO:sct:fslview t1.nii.gz t1_seg.nii.gz -l Red -b 0,1 -t 0.7 &
~~~

**Comment by jcohenadad** [comment]
@poquirion: a very useful feature would be to add "-mpi" to the field in the file version.txt, in case the user installs SCT with -mpi option.

**Comment by poquirion** [comment]
Comment @peristeri, make sure we do not add color to the file log.

**Comment by poquirion** [comment]
The verbose problem (getting double print) is absolutely weird. I appears when I lock a thread... I am simplifying the problem an opening a stack overflow question.

**Comment by poquirion** [comment]
Ouf after a panic stage, I manage to find the problem for the double log. It was in the distibute2mpi package.

**Comment by poquirion** [comment]
@jcohenadad,
It think the loggin part is good and ready for merging.
I still need to document usage and good practice, but it you don't know about it an continue working with the software you should no see any difference...
I am reseving the mpi part for an other issue. (the fact that is now an older mpi version that is shipper by default in

**Comment by jcohenadad** [comment]
tested version: 

there is a duplication of verbose (all lines starting with "INFO:sct:"). See:

~~~
~/sct_example_data/t2 $ sct_propseg -i t2.nii.gz -c t2
...
INFO:sct:Sucessfully generated the QC results in /Users/julien/qc_batch_processing/qc_results.json
Use the following command to see the results in a browser:
INFO:sct:Use the following command to see the results in a browser:
sct_qc -folder /Users/julien/qc_batch_processing/
INFO:sct:sct_qc -folder /Users/julien/qc_batch_processing/

Done! To view results, type:
INFO:sct:
Done! To view results, type:
fslview t2.nii.gz t2_seg.nii.gz -l Red -b 0,1 -t 0.7 &

INFO:sct:fslview t2.nii.gz t2_seg.nii.gz -l Red -b 0,1 -t 0.7 &
~~~

**Comment by poquirion** [comment]
@jcohenadad, the logging problem was a new issue.

@peristeri, the logging in qc was WRONG, that is SAD. Bad rating ;)

... its been fixed

File: scripts/sct_utils.py [hunk 1]
```diff
@@ -22,13 +22,104 @@
...
 # TODO: check if user has bash or t-schell for fsloutput definition
 
+"""
+Basic logging setup for the sct
+set SCT_LOG_LEVEL and SCT_LOG_FORMAT in ~/.sctrc to change the sct log
+format and level
+"""
+
+
+log = logging.getLogger('sct')
```
Comment (peristeri) [inline scripts/sct_utils.py]: I'm not sure about this. You have duplicated the contents of the function `start_stream_logger`.
I will try to refactor the initiation of the logger to make it more customizable.

Comment (jcohenadad) [reply scripts/sct_utils.py]: hi, are there any updates about this issue?

Comment (peristeri) [reply scripts/sct_utils.py]: I will submit a patch tonight. But I have not finished reviewing the PR.

Comment (poquirion) [reply scripts/sct_utils.py]: The `start_stream_logger` just does that, it start the logger, I get the logger at the root of `sct_utils` so we can use the `sct_utils.log.info()` that same way in all modules. 

Then, the `stream_handler` in initialise at the root  so its easier to pause the logging with  `pause_stream_logger`

I think is a good trade off between code usability and code "cleaniness".

Comment (jcohenadad) [reply scripts/sct_utils.py]: any update? has the issue been fixed? can we test?

Comment (peristeri) [reply scripts/sct_utils.py]: @jcohenadad So far, the code looks sound and for the first iteration, it works and can be tested.

Comment (jcohenadad) [reply scripts/sct_utils.py]: the issue I reported 12 days ago is still there. 

the following was tested on commit 1fdd46aea1831efe3e349944c0d311b33850a71b:
~~~
sct_propseg -i t2.nii.gz -c t2

--
Spinal Cord Toolbox (poq_issue_1409_log/1fdd46aea1831efe3e349944c0d311b33850a71b)
Running /Users/julien/code/sct/scripts/sct_propseg.py -i t2.nii.gz -c t2
Detecting the spinal cord using OptiC
Generate output files...
Finished. To view results, type:
fslview t2_int16.nii.gz &
isct_propseg -i "/Users/julien/sct_example_data/t2/t2.nii.gz" -t t2 -o "./" -verbose -init-centerline ./t2_centerline_optic.nii.gz -centerline-binary
Initialization - using given centerline
Total propagation length = 236.212 mm

Segmentation finished. To view results, type:
fslview /Users/julien/sct_example_data/t2/t2.nii.gz ./t2_seg.nii.gz &
Check consistency of segmentation...

Create temporary folder...
tmp.segmentation.nii.gz
Get dimensions of data...
64 x 320 x 320 x 1
Change orientation...
Generate output files...
WARNING: File tmp.segmentation_RPI.nii.gz already exists. Deleting it.
Finished. To view results, type:
fslview tmp.segmentation_RPI.nii.gz &
tmp.centerline.nii.gz
Get dimensions of data...
64 x 320 x 320 x 1
Change orientation...
Generate output files...
WARNING: File tmp.centerline_RPI.nii.gz already exists. Deleting it.
Finished. To view results, type:
fslview tmp.centerline_RPI.nii.gz &
Get data dimensions...
tmp.segmentation_RPI_c.nii.gz
Get dimensions of data...
64 x 320 x 320 x 1
Change orientation...
rm -f tmp.segmentation_RPI_c_LPI.nii.gz
Generate output files...
WARNING: File /Users/julien/sct_example_data/t2/t2_seg.nii.gz already exists. Deleting it.
Finished. To view results, type:
fslview /Users/julien/sct_example_data/t2/t2_seg.nii.gz &
Remove temporary files...
WARNING: File t2_seg.nii.gz already exists. Deleting it.
fatal: Not a git repository (or any of the parent directories): .git
Sucessfully generated the QC results in /Users/julien/qc_data/qc_results.json
INFO:sct:Sucessfully generated the QC results in /Users/julien/qc_data/qc_results.json
Use the following command to see the results in a browser:
INFO:sct:Use the following command to see the results in a browser:
sct_qc -folder /Users/julien/qc_data
INFO:sct:sct_qc -folder /Users/julien/qc_data
Done! To view results, type:
INFO:sct:Done! To view results, type:
fslview t2.nii.gz t2_seg.nii.gz -l Red -b 0,1 -t 0.7 &
INFO:sct:fslview t2.nii.gz t2_seg.nii.gz -l Red -b 0,1 -t 0.7 &
~~~

Comment (poquirion) [reply scripts/sct_utils.py]: Have you done a full install? It seems that I cannot reproduce the error here.

Comment (jcohenadad) [reply scripts/sct_utils.py]: i did not do a full install, as this was not mentioned at the beginning of the PR. In the future, please mention if full install is required (or partial, -m -b or -d depending on the case). Thanks. Testing it now

Comment (jcohenadad) [reply scripts/sct_utils.py]: ok, tested and I don't see the duplicate anymore. However, the text on the last line (see below) should appear in green as was the case before:
~~~
fslview t2.nii.gz t2_seg.nii.gz -l Red -b 0,1 -t 0.7 &
~~~

Comment (poquirion) [reply scripts/sct_utils.py]: I overefractor `sct_propseg` a bit here and transformed printv with color into a full sct.log.info(). I was desperate to fix the bug, that when I realized that I needed to do a `install_sct`...
I can revert to color.

**Comment by jcohenadad** [comment]
tested `batch_processing` with commit 1fdd46aea1831efe3e349944c0d311b33850a71b

still a few issues, but that could be addressed once merged. Few examples:
on `sct_register_multimodal`, these lines used to appear on the same row, now they appear as new lines:
~~~
Split destination volume...
0/128..
1/128..
2/128..
3/128..
4/128..
5/128..
6/128..
7/128..
8/128..
9/128..
10/128..
11/128..
12/128..
13/128..
14/128..
15/128..
16/128..
17/128..
18/128..
19/128..
20/128..
21/128..
22/128..
23/128..
24/128..
~~~

same comment with `sct_process_segmentation.py`
~~~
Remaining time: 00:03:42.84 (1/1600)                      
Remaining time: 00:02:49.16 (2/1600)                      
Remaining time: 00:02:21.21 (3/1600)                      
Remaining time: 00:02:21.40 (4/1600)                      
Remaining time: 00:02:13.92 (5/1600)                      
Remaining time: 00:02:11.14 (6/1600)                      
Remaining time: 00:02:05.66 (7/1600)                      
Remaining time: 00:02:07.84 (8/1600)                      
Remaining time: 00:02:07.53 (9/1600)
...
~~~
@poquirion: how can we refactor so that previous behaviour is kept the same?

another bigger issue: `sct_qc` crashes:
~~~
sct_qc -folder ~/qc_batch_processing

--
Spinal Cord Toolbox (poq_issue_1409_log/1fdd46aea1831efe3e349944c0d311b33850a71b)
Running /Users/julien/code/sct/scripts/sct_qc.py -folder /Users/julien/qc_batch_processing
Check folder existence...
QC viewer started on:
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_qc.py", line 68, in <module>
    sct.log.info('http://127.0.0.1:8888', type='info')
  File "/Users/julien/code/sct/python/lib/python2.7/logging/__init__.py", line 1167, in info
    self._log(INFO, msg, args, **kwargs)
TypeError: _log() got an unexpected keyword argument 'type'
~~~
@peristeri: please fix

File: scripts/sct_utils.py (hunk 1 above)
Comment (poquirion) [reply scripts/sct_utils.py]: Reverted to color in latest commit 2403aff

**Pull request merged**